components/calendar/src/any_calendar.rs

```diff
@@ -1026,46 +877,71 @@ impl AnyCalendarKind {
-            let lang = prefs.locale_preferences.language();
-            if lang == language!("th") {
-                Self::Buddhist
-            } else if lang == language!("sa") {
-                Self::HijriUmmAlQura
-            } else if lang == language!("af") || lang == language!("ir") {
-                Self::Persian
-            } else {
-                Self::Gregorian
+impl TryFrom<CalendarAlgorithm> for AnyCalendarKind {
+    type Error = ();
+    fn try_from(v: CalendarAlgorithm) -> Result<Self, Self::Error> {
+        use CalendarAlgorithm::*;
+        match v {
+            Buddhist => Ok(AnyCalendarKind::Buddhist),
+            Chinese => Ok(AnyCalendarKind::Chinese),
+            Coptic => Ok(AnyCalendarKind::Coptic),
+            Dangi => Ok(AnyCalendarKind::Dangi),
+            Ethioaa => Ok(AnyCalendarKind::EthiopianAmeteAlem),
+            Ethiopic => Ok(AnyCalendarKind::Ethiopian),
+            Gregory => Ok(AnyCalendarKind::Gregorian),
+            Hebrew => Ok(AnyCalendarKind::Hebrew),
+            Indian => Ok(AnyCalendarKind::Indian),
+            Hijri(None) => Ok(AnyCalendarKind::HijriObservational),
+            Hijri(Some(HijriCalendarAlgorithm::Umalqura)) => Ok(AnyCalendarKind::HijriUmmAlQura),
+            Hijri(Some(HijriCalendarAlgorithm::Tbla)) => Ok(AnyCalendarKind::HijriTabular),
+            Hijri(Some(HijriCalendarAlgorithm::Civil)) => Ok(AnyCalendarKind::HijriCivil),
+            Hijri(Some(HijriCalendarAlgorithm::Rgsa)) => Err(()),
+            Iso8601 => Ok(AnyCalendarKind::Iso),
+            Japanese => Ok(AnyCalendarKind::Japanese),
+            Persian => Ok(AnyCalendarKind::Persian),
+            Roc => Ok(AnyCalendarKind::Roc),
+            _ => {
+                debug_assert!(false, "unknown calendar algorithm {v:?}");
+                Err(())
             }
         }
     }
 }
 
+impl From<AnyCalendarKind> for CalendarAlgorithm {
+    fn from(v: AnyCalendarKind) -> Self {
+        use AnyCalendarKind::*;
+        match v {
+            Buddhist => CalendarAlgorithm::Buddhist,
+            Chinese => CalendarAlgorithm::Chinese,
+            Coptic => CalendarAlgorithm::Coptic,
+            Dangi => CalendarAlgorithm::Dangi,
+            EthiopianAmeteAlem => CalendarAlgorithm::Ethioaa,
+            Ethiopian => CalendarAlgorithm::Ethiopic,
+            Gregorian => CalendarAlgorithm::Gregory,
+            Hebrew => CalendarAlgorithm::Hebrew,
+            Indian => CalendarAlgorithm::Indian,
+            HijriObservational => CalendarAlgorithm::Hijri(None),
+            HijriUmmAlQura => CalendarAlgorithm::Hijri(Some(HijriCalendarAlgorithm::Umalqura)),
+            HijriTabular => CalendarAlgorithm::Hijri(Some(HijriCalendarAlgorithm::Tbla)),
+            HijriCivil => CalendarAlgorithm::Hijri(Some(HijriCalendarAlgorithm::Civil)),
+            Iso => CalendarAlgorithm::Iso8601,
+            Japanese => CalendarAlgorithm::Japanese,
+            JapaneseExtended => CalendarAlgorithm::Japanese,
```

Comment:
Thought: This is lossy.

I guess we can fix this when https://unicode-org.atlassian.net/browse/CLDR-15669 gets implemented.

Reply:
https://github.com/unicode-org/icu4x/issues/4889

Reply:
Yeah, we should fix #4889.